test/functional/shada/shada_spec.lua

```diff
@@ -262,4 +262,19 @@ describe('ShaDa support code', function()
        exc_exec('wshada'))
     meths.set_option('shada', '')
   end)
+
+  it('does not store unlisted buffer', function()
+    local fname = funcs.getcwd() .. '/file'
+    meths.set_var('__fname', fname)
+    nvim_command('silent! edit `=__fname`')
+    curbufmeths.set_option('buflisted', false)
+    nvim_command('wshada! ' .. shada_fname)
+
+    eq({}, find_file(fname))
+
+    -- Set 'buflisted', then check again.
+    curbufmeths.set_option('buflisted', true)
+    nvim_command('wshada! ' .. shada_fname)
+    eq({[7]=1, [8]=1, [10]=1}, find_file(fname))
```

Comment:
because this depends on the filesystem it can be racy. try wrapping this in `retry()` 
```suggestion
    retry(nil, 4000, function()
      eq({[7]=1, [8]=1, [10]=1}, find_file(fname)) end)
```